### PR TITLE
Remove duplication of archive file on load.

### PR DESCRIPTION
**Pull request by DrSuperGood**

### CODE REVIEW

**Comment by Frotty** [comment]
Hey, thanks a lot for the PR.
My only issue is formatting right now, which makes the diff show the entire JMPQEditor as changed, same for the tests.
If you don't mind it would be nice if you would keep the original formatting or at least increase your line width to 160.
Lastly maybe provide a constructor that passes FORCEV0 by default, since users might not know what to use.

e: I can also do this when merging ofc.

**Comment by DrSuperGood** [comment]
> My only issue is formatting right now, which makes the diff show the entire JMPQEditor as changed, same for the tests. If you don't mind it would be nice if you would keep the original formatting or at least increase your line width to 160.

Ill try formatting single element in future.

> Lastly maybe provide a constructor that passes FORCEV0 by default, since users might not know what to use.

This might affect the ability to load StarCraft II maps as forcing version 0 will prevent usage of the decompression used in newer formats as well as parsing the newer header structures which may or may not prevent loading the MPQ archive. This is why MPQEdit needs one to open Spazzled maps specifically with a similar option enabled as doing so disables all version awareness and mirrors WC3 mechanics more closely.

In case there is any confusion. Without any open options it will load WC3 maps perfectly. Just maps that have deliberately been damaged, eg by Spazzler, need the force version 0 open option as they corrupt the version field and add user data headers to try and trick MPQ applications into thinking they are a newer MPQ format version than the version 0 they are. Only reason Warcraft III can even load such maps is because Blizzard never updated the MPQ component code so it is completely unaware of anything not format version 0 and will even ignore the format version field entirely (instead of refusing to load something a newer version than it was designed for).

Is anything not clear with the documentation? The FORCE_V0 option explicitly mentions that it is needed to load deliberately damaged MPQ archives. Would it be a good idea to bring in WC3 terminology such as "protected maps"? Or maybe make it more clear the option may be required in the constructor documentation?

**Comment by Frotty** [comment]
btw you can select in eclipse to only format edited lines.

I worked around Spazzler before already by ignoring everything before the correct mpq header, I suppose this is the userdata part?

The documentation is fine, though I doubt that common users have any more idea about "mpq-version 0" than "protection".
There should simply be an option that usually works, especially becazse this would break backwards compatibility with projects using jmpq, e.g. wurstscript, which isn't a great idea.

**Comment by DrSuperGood** [comment]
I have reformatted the affected files to 160 file lines. I will only format lines I change in future.

> I worked around Spazzler before already by ignoring everything before the correct mpq header, I suppose this is the userdata part?

Yes it is. User data was added after MPQ version 0 and comes in the form of chunks that redirect to the actual MPQ header. Apparently it is used by StarCraft II, however it is unclear if this is for custom maps or for the now removed main data archives (replaced with CASC).

Spazzler works by offsetting the MPQ header to 1024 bytes and replacing the 512 byte header with a nonsense user data header which pointed at a more nonsense MPQ header inserted after it. This is on top of corrupting fields such as archive size and block size in the actual header. Warcraft III still opens the archive as it skips the user data header at 512 bytes (not recognized) and detects the correct one at 1024 bytes. A general purpose MPQ library built to support up to MPQ format version 3 (StarCraft II) reads the nonsense user data header, jumps to the nonsense MPQ header and then fails. The only sensible way for such a library to load such MPQ archives is to forcefully instruct it to follow MPQ format version 0 standards.

> There should simply be an option that usually works, especially becazse this would break backwards compatibility with projects using jmpq, e.g. wurstscript, which isn't a great idea.

How often does wurstscript need to manipulate a spazzled map? I thought it was used in map creation and not post processing. In any case I agree that backwards compatibility is important.

I could restore the original constructors so that they imply FORCE_V0, deprecate them (so that they are not used by future projects) and then supply the new constructors next to them for future use. This will allow existing projects to still open Spazzled maps, however it will mean that trying to open any non format version 0 archive (eg a SC2 map) will not work. That is if opening non format version 0 worked in the first place as the code was not even trying to parse header content beyond version 0 before these commits. This is probably the best solution if spazzled and damaged maps are already a common use case for this library.

**Comment by Frotty** [comment]
Okay sounds good. However the formatting only got worse. Please see the diff for yourself
https://github.com/inwc3/JMPQ3/pull/12/files

It's not crucial though.

**Comment by DrSuperGood** [comment]
Hopefully have fixed the formatting. It seems it was using 4 wide tabs instead of 4 spaces, which might be why it said every line was deleted and new. Might have to flatten these commits though for correct metrics and useful history.

**Comment by Frotty** [comment]
Thanks a lot. Github supports automatic squashing on PR merge 👍

File: src/main/java/systems/crigges/jmpq3/JMpqEditor.java [hunk 1]
```diff
@@ -266,27 +266,50 @@ public JMpqEditor(Path mpqArchive, MPQOpenOption... openOptions) throws JMpqExce
...
-	 *             if mpqArchive is not a file or does not exist and READ_ONLY
-	 *             option is specified.
 	 * @throws JMpqException
 	 *             if mpq is damaged or not supported.
 	 */
-	public JMpqEditor(File mpqArchive, MPQOpenOption... openOptions) throws IOException {
-		this(mpqArchive.toPath(), openOptions);
+	@Deprecated
+	public JMpqEditor(File mpqArchive) throws IOException {
+		this(mpqArchive.toPath(), MPQOpenOption.FORCE_V0);
```
Comment (Cokemonkey11) [inline src/main/java/systems/crigges/jmpq3/JMpqEditor.java]: Using an enum I don't think it's possible to open `FORCE_V0` and `READONLY` at the same time, is that right?

Probably worth revisiting.

**Comment by Cokemonkey11** [comment]
regarding whitespace change: FYI you can use https://github.com/inwc3/JMPQ3/pull/12/files?w=1 `?w=1`

**Pull request merged**